lib/Dialect/TT/IR/TTOpsTypes.cpp

```diff
@@ -513,8 +511,79 @@ MemorySpace LayoutAttr::getMemorySpace() const {
       .getValue();
 }
 
-mlir::AffineMap LayoutAttr::projectOnto(ArrayRef<int64_t> logicalTensorShape,
-                                        GridAttr grid) const {
+// Returns shape of the tensor after tilization is applied.
+llvm::SmallVector<int64_t>
+LayoutAttr::getTiledShape(llvm::ArrayRef<int64_t> tensorShape) const {
+  assert(isTiled() && "Expected a tiled layout");
+
+  auto tileLinearMap = getTileLinearMap();
+
+  mlir::SmallVector<mlir::AffineExpr> tensorShapeExprs(
+      llvm::map_range(tensorShape, [context = getContext()](std::int64_t e) {
+        return getAffineConstantExpr(e - 1, context);
+      }));
+
+  mlir::SmallVector<std::int64_t> tiledShape(tileLinearMap.getNumResults());
+  for (unsigned i = 0; i < tileLinearMap.getNumResults(); ++i) {
+    mlir::AffineExpr expr = tileLinearMap.getResult(i);
+    mlir::AffineExpr constantExpr = expr.replaceDims(tensorShapeExprs);
+    tiledShape[i] =
+        llvm::cast<mlir::AffineConstantExpr>(constantExpr).getValue() + 1;
+  }
```

Comment:
All of this code can be written more succinctly, this was before I knew about this special overload of affine map [`compose`](https://mlir.llvm.org/doxygen/classmlir_1_1AffineMap.html#a7c205e3ff1701beab13d721c11031ed2), I think somewhere else in this file is where this routine came from.  Anyway, I think this whole function can become:
```c++
auto tileLinearMap = getTileLinearMap();
return tileLinearMap.compose(tensorShape);
```

We should probably eventually fixup the other places that have used this routine, but that's outside the scope of your PR.

Reply:
Yeah this is a much nicer way. Moved `evalShape` to Utils.h.